### ISSUE TITLE
2.0.6: KCL in recurring "Sleeping..." state while events kept flowing into the source Kinesis stream

### ISSUE THREAD

**Issue by arthur78** [snowplow/snowplow-elasticsearch-loader#236]
Hi guys,

We came across a situation when the loader app at some moment went into the recurring "Sleeping" state, while the events were keep coming into the source Kinesis stream.

Here's the excerpt from the app's log:

2022-06-25 08:08:28:447 +0000 [RecordProcessor-0000] INFO com.snowplowanalytics.stream.loader.clients.ElasticsearchBulkSender - Emitted 7 records to Elasticsearch
2022-06-25 08:08:28:452 +0000 [RecordProcessor-0000] INFO com.amazonaws.services.kinesis.clientlibrary.lib.worker.SequenceNumberValidator - Validated sequence number 49630759577339465381993565886722464592375289110455123970 with shard id shardId-000000000000
2022-06-25 08:08:58:385 +0000 [main] INFO com.amazonaws.services.kinesis.clientlibrary.lib.worker.Worker - Current stream shard assignments: shardId-000000000000
2022-06-25 08:08:58:385 +0000 [main] INFO com.amazonaws.services.kinesis.clientlibrary.lib.worker.Worker - Sleeping ...
2022-06-25 08:09:13:744 +0000 [pool-8-thread-1] INFO com.amazonaws.services.kinesis.leases.impl.LeaseCleanupManager - Number of pending leases to clean before the scan : 0
2022-06-25 08:09:18:255 +0000 [pool-6-thread-1] INFO com.amazonaws.services.kinesis.clientlibrary.lib.worker.DeterministicShuffleShardSyncLeaderDecider - Elected leaders: 786870d43f288b98:-4e439256:1818d0792db:-8000
2022-06-25 08:09:59:395 +0000 [main] INFO com.amazonaws.services.kinesis.clientlibrary.lib.worker.Worker - Current stream shard assignments: shardId-000000000000
2022-06-25 08:09:59:395 +0000 [main] INFO com.amazonaws.services.kinesis.clientlibrary.lib.worker.Worker - Sleeping ...
2022-06-25 08:10:13:744 +0000 [pool-8-thread-1] INFO com.amazonaws.services.kinesis.leases.impl.LeaseCleanupManager - Number of pending leases to clean before the scan : 0
2022-06-25 08:10:24:738 +0000 [pool-7-thread-1] INFO com.amazonaws.services.kinesis.clientlibrary.lib.worker.PeriodicShardSyncManager - Skipping shard sync due to the reason - Hash range is complete.
2022-06-25 08:11:00:405 +0000 [main] INFO com.amazonaws.services.kinesis.clientlibrary.lib.worker.Worker - Current stream shard assignments: shardId-000000000000
2022-06-25 08:11:00:405 +0000 [main] INFO com.amazonaws.services.kinesis.clientlibrary.lib.worker.Worker - Sleeping ...
2022-06-25 08:11:13:744 +0000 [pool-8-thread-1] INFO com.amazonaws.services.kinesis.leases.impl.LeaseCleanupManager - Number of pending leases to clean before the scan : 0
2022-06-25 08:12:01:421 +0000 [main] INFO com.amazonaws.services.kinesis.clientlibrary.lib.worker.Worker - Current stream shard assignments: shardId-000000000000
2022-06-25 08:12:01:421 +0000 [main] INFO com.amazonaws.services.kinesis.clientlibrary.lib.worker.Worker - Sleeping ...
2022-06-25 08:12:13:747 +0000 [pool-8-thread-1] INFO com.amazonaws.services.kinesis.leases.impl.LeaseCleanupManager - Number of pending leases to clean before the scan : 0
2022-06-25 08:12:24:742 +0000 [pool-7-thread-1] INFO com.amazonaws.services.kinesis.clientlibrary.lib.worker.PeriodicShardSyncManager - Skipping shard sync due to the reason - Hash range is complete.
2022-06-25 08:13:02:432 +0000 [main] INFO com.amazonaws.services.kinesis.clientlibrary.lib.worker.Worker - Current stream shard assignments: shardId-000000000000
2022-06-25 08:13:02:432 +0000 [main] INFO com.amazonaws.services.kinesis.clientlibrary.lib.worker.Worker - Sleeping ...
2022-06-25 08:13:13:744 +0000 [pool-8-thread-1] INFO com.amazonaws.services.kinesis.leases.impl.LeaseCleanupManager - Number of pending leases to clean before the scan : 0
2022-06-25 08:14:03:448 +0000 [main] INFO com.amazonaws.services.kinesis.clientlibrary.lib.worker.Worker - Current stream shard assignments: shardId-000000000000
2022-06-25 08:14:03:448 +0000 [main] INFO com.amazonaws.services.kinesis.clientlibrary.lib.worker.Worker - Sleeping ...
2022-06-25 08:14:13:744 +0000 [pool-8-thread-1] INFO com.amazonaws.services.kinesis.leases.impl.LeaseCleanupManager - Number of pending leases to clean before the scan : 0

---
We don't see any WARN/ERROR messages in the logs. Such the logging pattern was repeated until we restarted the app today.
After the restart, it has resumed reading the events. 

We made sure from the source stream metrics that events continued to flow into it.

Maybe you have an idea what could be in here?

Thanks.

**Comment by istreeter** [comment]
Hi @arthur78 thanks for reporting this.  I'm investigating a separate issue for elasticsearch loader, which might (might!!) be related to this issue.

By any chance, do you have the logs going back just a little bit earlier?  Your first log message you shared is from `2022-06-25 08:08:28:447`, but it would be helpful to see logs from where it processed the previous batch of events.

**Comment by arthur78** [comment]
Hi @istreeter ,

In our case, the above situation starts to take place after the ES cluster responds sporadically with 403 Forbidden on a write request from the shard's RecordProcessor – that blocks the processor, resulting in no next events are being read from the shard.

Please find some more details at https://discourse.snowplow.io/t/sleeping-kinesis-stream-shards-with-latest-es-loaders/7122

(Also, those unexpected sporadic 403 Forbidden responses from the ES cluster are not observed in the logs of the older version (0.10.1) of the ES loader we are keep using. However, the flow of events in its source Kinesis stream is not that intensive.)

**Comment by arthur78** [comment]
@istreeter , please let me know if you need the Debug-level logs.

**Comment by istreeter** [comment]
Thanks again for sharing more details.  Issue #238 describes some issues I've seen, and a few suggested fixes, although they are partly guesses.  It _might_ be the same issue you're describing here, but it's hard to tell.

I will update this thread once there is a new loader version to test.

**Comment by istreeter** [comment]
Hi @arthur78 I have just released Elasticsearch Loader version 2.0.7 which might solve your issue as it addresses #238.  Upgrading is as simple as pulling the newer docker image

If you find the problem is fixed with the newer version, please could you let me know?

**Comment by arthur78** [comment]
Sure, @istreeter, thank you! (I plan to start testing it asap.)

**Comment by arthur78** [comment]
Hi @istreeter! Surprisingly, the 2.0.7 loader app instances seem to fall into the "Sleeping" scenario even faster than in the previous 2.0.6 version. 

We have 9 loaders running (5 of them are running in parallel on two separate AWS EC2s). If you like, we can collect some debug logs for you, e.g. since the loader app start and until it falls "sleeping" on one or more of the source stream's shards. Please let me know.

**Comment by istreeter** [comment]
Hi @arthur78, thanks for following up with feedback, and sorry to hear the newest version didn't improve anything for you.

There's a couple of things I would be interested to know:
- How many kinesis shards do you have in your input (enriched) stream?  This is interesting to me, because the shards will be shared out between the 9 loaders.
- Do you ever scale up or scale down the number of kinesis shards, or is it a consistent number?
- Which EC2 instance type are you using?  I'm interested in how much total memory and cpu you have available.
- Do you limit how much memory the docker image or JVM is allowed to use?

> We have 9 loaders running (5 of them are running in parallel on two separate AWS EC2s

That is quite an unusual setup; normally I would expect to see 1 loader per EC2 instance.  The loader is designed to use all the cores and memory available on the instance, and it is designed to spawn several "workers" internally to handle as many shards as needed.  So there should be no benefit of running more than 1 loader on the same node.  I wonder if this is a case of over-allocating memory caused by running too many loaders.  That's a bit of a guess, but we've had issues before with kinesis apps doing strange things when there's not enough memory available.

**Comment by arthur78** [comment]
Hi @istreeter ,

We have 9 source Kinesis streams. The numbers of shards in the streams are different, from 1 and up to 20 shards in a stream. Yes, we did up-sharding a couple of times, but only in Prod. (The new 2.0.x loaders are only in Staging now.)

5 of the loaders run in parallel on two separate EC2s, i.e. 2 loader instances per stream: first loader instance on first EC2, second loader instance – on second EC2. The remaining 4 source streams are loaded from a third EC2.

Please note that we don't have the problem described in that issue in the old 0.10.1 version of the loader. Just in case, I'm repeating the forum link https://discourse.snowplow.io/t/sleeping-kinesis-stream-shards-with-latest-es-loaders/7122 where I provided more details about it, including findings by AWS support.

The first two mentioned Staging EC2s have 2 CPUs and 4 GB RAM (1+ GB free), the third EC2 has 2 CPUs and 8 GB RAM (~6 GB free). The load average is normally small on all three EC2s (~0.2). We do not change memory limits for JVM.

**Comment by istreeter** [comment]
Hi @arthur78 thanks is helpful, and it makes sense now why you are running multiple loaders on the same EC2.

> If you like, we can collect some debug logs for you, e.g. since the loader app start and until it falls "sleeping" on one or more of the source stream's shards. Please let me know.

Please can I take you up on this offer?  It would be very helpful to see logs starting shortly before the error when everything is working OK, through to when it stops processing.

I read through [your explanation in the discourse post](https://discourse.snowplow.io/t/sleeping-kinesis-stream-shards-with-latest-es-loaders/7122).  Do I understand right that the spike in the lower graph corresponds to when you reboot the loader?  I agree with the AWS engineers that this looks like a Snowplow problem, not an AWS problem!!!  So I hope we can track down the problem and fix it for you.

**Comment by arthur78** [comment]
Hi @istreeter ,

Sure, I've enabled the debug-level logging on one of the loaders. Tell me please, is there a preferred method of sharing the log file?

Yes, your understanding regarding the loader reboot time is correct.

**Comment by istreeter** [comment]
@arthur78 how about share it in [the discourse thread on this same topic](https://discourse.snowplow.io/t/sleeping-kinesis-stream-shards-with-latest-es-loaders/7122).  If you hit "reply" then there is an option in the toolbar to upload at attachment.

Or if for some reason that doesn't work, then I can share my email address.

**Comment by arthur78** [comment]
Hi @istreeter , 

Can you please take it from this GDrive [link](https://drive.google.com/file/d/15OJ-JL8kwOFmhpAnj3m7e2wJPvlEsnFx/view)?

**Comment by arthur78** [comment]
And here's the [password](https://onetimesecret.com/secret/epzqyl6o5posa2gez6s3f34bpq82o1g) for zip.

**Comment by istreeter** [comment]
Thank you @arthur78 your logs were very helpful.  I have opened two new issues #241 and #242 to address the problems I saw in the logs.  I hope to get these fixes released soon as version 2.0.8.

If you're interested in testing the fixes early, then there is a release candidate available already; just pull docker image version 2.0.8-rc1.

**Comment by arthur78** [comment]
Thank you @istreeter, good luck in fixing this!

**Comment by istreeter** [comment]
Hi @arthur78 we released version 2.0.8 of elasticsearch loader a few days ago, and on our pipelines it is proving to be very stable so far.  I would be interested to hear if you also find it to be more stable than previous releases?

**Comment by arthur78** [comment]
Hi @istreeter , we run 2.0.8 in our staging for 22 hours now, and so far no "sleeping" alarms, which already looks promising! Let me keep an eye on it for a couple of days more and I get back with another update. Thank you!

**Comment by arthur78** [comment]
Hi @istreeter , we run this version for more than a week now, in our prod as well. It looks good so far. Thank you very much!
